qiskit/transpiler/passes/crosstalk_adaptive_schedule.py

```diff
@@ -0,0 +1,694 @@
+    """
+    Crosstalk mitigation through adaptive instruction scheduling.
+    """
+    def __init__(self, backend_prop, crosstalk_prop, weight_factor=0.5, measured_qubits=None):
+        """
+        CrosstalkAdaptiveSchedule initializer.
+
+        Args:
+            backend_prop (BackendProperties): backend properties object
+            crosstalk_prop (dict): crosstalk properties object
+                crosstalk_prop[g1][g2] specifies the conditional error rate of
+                g1 when g1 and g2 are executed simultaneously.
+                g1 should be a two-qubit tuple of the form (x,y) where x and y are physical
+                qubit ids. g2 can be either two-qubit tuple (x,y) or single-qubit tuple (x).
+                We currently ignore crosstalk between pairs of single-qubit gates.
+                Gate pairs which are not specified are assumed to be crosstalk free.
+
+                Example: crosstalk_prop = {(0, 1) : {(2, 3) : 0.2, (2) : 0.15},
+                                           (4, 5) : {(2, 3) : 0.1},
+                                           (2, 3) : {(0, 1) : 0.05, (4, 5): 0.05}}
+                The keys of the crosstalk_prop are tuples for ordered tuples for CX gates
+                e.g., (0, 1) corresponding to CX 0, 1 in the hardware.
+                Each key has an associated value dict which specifies the conditional error rates
+                with nearby gates e.g., (0, 1) : {(2, 3) : 0.2, (2) : 0.15} means that
+                CNOT 0, 1 has an error rate of 0.2 when it is executed in parallel with CNOT 2,3
+                and an error rate of 0.15 when it is executed in parallel with a single qubit
+                gate on qubit 2.
+            weight_factor (float): weight of gate error/crosstalk terms in the objective
+                weight_factor*fidelities + (1-weight_factor)*decoherence errors.
+                Weight can be varied from 0 to 1, with 0 meaning that only decoherence
+                errors are optimized and 1 meaning that only crosstalk errors are optimized.
+                weight_factor should be tuned per application to get the best results.
+            measured_qubits (list): a list of qubits that will be measured in a particular circuit.
+                This arg need not be specified for circuits which already include measure gates.
+                The arg is useful when a subsequent module such as state_tomography_circuits
+                inserts the measure gates. If CrosstalkAdaptiveSchedule is made aware of those
+                measurements, it is included in the optimization.
+        Raises:
+            ImportError: if unable to import z3 solver
+
+        """
+        super().__init__()
+        if not Z3_AVAIL:
+            raise ImportError('z3-solver is required to use CrosstalkAdaptiveSchedule')
+        self.backend_prop = backend_prop
+        self.crosstalk_prop = crosstalk_prop
+        self.gate_id = {}
```

Comment:
Through inspection, and because I also know the `backend` methods you used return time in `s`, it looks like you're using nanoseconds as the units for `dur` and `time`. Furthermore, the values are truncated at nanoseconds. If another developer looks at this code (or you, months/years from now), would that be apparent? Can you think of a way to make it obvious?

Reply:
Added a comment.